File: CSE262/p2/src/python/slang.py
import sys
import slang_scanner
import slang_scanner_xml


def getFromStdin(prompt):
    """Read a line from standard input (typically the keyboard).  If nothing is
    provided, return an empty string."""
    try:
        return input(prompt).rstrip()
    except EOFError:
        # we print a newline, because this is going to cause an exit, and it
        # will be a bit more uniform if all ways of exiting have a newline
        # before the shell is back in control
        print("")
        return ""


def getFile(filename):
    """Read a file and return its contents as a single string"""
    try:
        source_file = open(filename, "r")
        code = source_file.read()
        source_file.close()
        return code
    except FileNotFoundError:
        print("Error: file not found", file=sys.stderr)
        return ""


def printHelp():
    """Print a help message"""
    print("slang -- An interpreter for a Scheme-like language (Python version)")
    print("  Usage: slang [mode] [filename]")
    print("    * If no filename is given, a REPL will run")
    print("    * If a filename is given, the entire file will be loaded and evaluated")
    print("  Modes:")
    print("    -help             Display this message and exit")
    print("    -scan             Turn slang code into XML tokens")


def main(args):
    """Run the Scheme interpreter"""
    # Parse the command-line arguments.  Make sure exactly one valid mode is
    # given, and at most one filename
    filename, mode, numModes, numFiles = "", "", 0, 0
    for a in args:
        if a in ["-help", "-scan"]:
            mode = a
            numModes += 1
        else:
            filename = a
            numFiles += 1
    if numModes != 1 or numFiles > 1 or mode == "-help":
        return printHelp()

    # Run the REPL loop, but only once if we have a valid filename
    while True:
        # Get some code
        codeToRun = ""
        if filename != "":
            codeToRun = getFile(filename)
        else:
            codeToRun = getFromStdin(":> ")
            if codeToRun == "":
                break

        # SCAN mode
        if mode == "-scan":
            tokens = slang_scanner.scanTokens(codeToRun)
            print(slang_scanner_xml.xmlToString(tokens))

        # exit if we just processed a file
        if filename != "":
            break


# In python, this is how we get main() to run when we invoke this program via
# `python3 slang.py ...`
if __name__ == "__main__":
    main(sys.argv[1:])

File: CSE262/p2/src/java/app/src/main/java/edu/lehigh/cse262/slang/Slang.java
package edu.lehigh.cse262.slang;

import edu.lehigh.cse262.slang.Scanner.Scanner;
import edu.lehigh.cse262.slang.Scanner.XmlTokenWriter;
import edu.lehigh.cse262.slang.Scanner.Scanner.ScanError;

/** Slang is the entry point into our interpreter */
public class Slang {
    public static void main(String[] args) {
        // Get the command-line arguments. If help is requested or needed,
        // print help and exit immediately.
        var parsedArgs = new Args(args);
        if (parsedArgs.mode == Args.Modes.HELP) {
            Args.printHelp();
            return;
        }

        // Create a SourceLoader and use it to start loading code and scanning
        // it. When given a file, we run exactly one iteration. Otherwise, we
        // run until we get an empty string of text from the sourceLoader.
        //
        // [CSE 262] You should be sure to understand this try-with-resources
        // syntax.
        try (var sourceLoader = new SourceLoader()) {
            do {
                // Get the code to run, break if no code is available
                String codeToRun;
                if (!parsedArgs.fileName.equals("")) {
                    codeToRun = sourceLoader.getFile(parsedArgs.fileName);
                } else {
                    codeToRun = sourceLoader.getFromStdin(":> ");
                    if (codeToRun.equals(""))
                        break;
                }

                // SCAN mode: read the source code, turn it into tokens, print
                // the tokens as XML.
                if (parsedArgs.mode == Args.Modes.SCAN) {
                    try {
                        var tokens = new Scanner().scanTokens(codeToRun);
                        var writer = new XmlTokenWriter();
                        writer.writeXmlToStream(tokens, System.out);
                    } catch (ScanError se) {
                        System.out.println(se.getMessage());
                    }
                }
            } while (parsedArgs.fileName.equals(""));
        }
    }
}
